### ISSUE TITLE
[Feature Request?] Context store for easily sharing data through machine hierarchy

### ISSUE THREAD

**Issue by jonlaing** [davidkpiano/xstate#366]
I'm working on a pretty complex app with many invoked machines. After a while, it gets kind of hard to keep track of where you've properly passed down context from a parent machine into an invoked one. As the app has gotten more complex, tracking down these errors has become more of a slog.

As an example, let say I have an ecommerce app, and I need to keep my Order up to date. So I obviously have a browse machine, a cart machine, a checkout machine, a credit card checkout machine, a paypal checkout machine, etc etc. Not only do each one of those need to keep track of the Order, but each one of them can make changes to it.

This was a problem for React a while ago, where you'd have to pass down a single prop way down the structure, and it got unwieldy fast. Redux was one solution, but also there are React Contexts which are more lightweight.

I'm proposing something like this:

```typescript
const BMachine = Machine({
  /* ... */
  states: {
    c: {
      onEntry: 'changeThing'
    }
  }
}, {
  actions: {
    changeThing: assignStore({
      thing: (ctx, evt, store) => store.thing + ' is the best'
    })
  }
});

Machine({
  /* ... */
  states: {
    a: {
      invoke: {
        src: 'b',
        shareStore: true
      }
  }.
  {
    services: {
        b: BMachine
    }
  }
}).withStore({ thing: 'my value' });
```

It's kind of a half baked idea, but I wanted to get thoughts on it. Maybe there's already an implementation of something like this that I missed, or maybe it's already in the pipeline, or maybe it's completely against the principles of the project. Figured I'd put it out there regardless.

Looking at it now, I guess I could just make my own store and inject it into my actions and what not. I've actually seen implementations of things like that. I just thought it might be good to add it to the official project. Let me know what y'all think.

**Comment by flowt-au** [comment]
I am interested in this pattern also and am in the very early stages of experimenting with it, so this comment is more theoretical than applied, at this stage. I do not use React so maybe this wont work. 

My approach is to encapsulate each of the machines mentioned in their own microservice and use messages to communicate between services (client side). ie the standard microservice pattern whereby the state and UI is hidden inside the service and each service responds to messages of interest to update its state and hence the UI. So, there are no "child" states as such. Each machine's state change emits a message (eg Order quantity went from 1 to 2) and which ever services are interested in order quantity changes hear that message and update their own states. 

I am playing with Seneca for browser [https://github.com/voxgig/seneca-browser](url) and am yet to start using XState as the state manager within each service. The UI is then also internal to the service. Theoretically one could use plain JS, or Vue, or jQuery, and I assume React? or any other UI library and indeed could have different libraries for each service which could be useful when migrating from one library to another. 

Anyway, just my thoughts. I guess I am questioning whether child states are the way to go here or whether the issue could be resolved at a higher level of abstraction. I would love to hear thoughts since I have not built this experiment fully yet and it could be a complete disaster in practice.

Cheers,
Murray

**Comment by flowt-au** [comment]
PS: to push the concept even further, by decoupling all the elements of an ecommerce site into microservices client side, each microservice could even use different state managers internally (or plain JS for really simple services) because none of them need to know about the others. They just emit and listen for messages  (which can be broadcast, subscribe, request, sync, async, etc as required)

**Comment by jonlaing** [comment]
@flowt-au I mean, that's essentially how everything is working now. The complication is that as my app has gotten more complex, the individual "services" (machines) may or may not have disparate context. For instance, if the user deletes an item from their cart, then my checkout needs to be updated (that's not really how it works in my actual app, but I couldn't think of a better contrived example). 

The current way I do this is by having a main machine that handles the broad strokes of the state that passes down context to its children, which pass it down to their children, etc via the `data` property of the `invoke`. Then each machine passes its changes back up the line via `sendParent` and they get merged.

This works, but as a developer, having to keep it in my head when and where I've merged my state manually is getting difficult, and it's led to times where one of my machines has the most current order, but another has an outdated version, and it causes bugs. 

Just as an example, my machine hierarchy goes something like this:

```
                                     Main
                       _________|_________
                       |                                   |
                   Cart                   [Bunch of other machines]
         _______|______
         |                            |
   Checkout         SelectCustomer
   ___|__
   |           | 
Card    Cash
```

`Card` and `Cash` need to know about the current customer, which is selected in `SelectCustomer`. So I have to send a message to `Cart` which updates its context, and then passes the update down the chain. Likewise, `Checkout`, `Card`, and `Cash` can all update the the cart object, which then has to be passed up the line all the way to the `Main` machine so that it can be displayed in the UI. The shopping cart object is kind of "global", but I'm constantly having to pass around versions of it that disagree and manually sync them.

In my mind, a machine's context shouldn't matter to the rest of the world, so if I need "global" data, it should be handled in another way.

What I've started to do, while this gets hashed out here, is to make a custom data store (just a JS class with a private data object and a way to listen for changes), and including in my actions and services manually. I just started doing this, so at the moment, it's a naive approach. It would be convenient to have my store as part of the machine somehow, but maybe this approach is fine.

**TL;DR** My main challenge is syncing all of the messages and updates from all the different machines. It's doable, but a very manual process that's prone to breaking due to developer error. I'm looking for a way have "global" context so I don't have to manually manage this.

PS: The UI isn't really an important part of this proposal. I was just using `React.Context` as an example of a similar problem solved in a similar way to what I'm proposing.

**Comment by flowt-au** [comment]
Yep, I understand what you are saying and it is different to what I was proposing in my theoretical construct. The difference is that each of the services (Cart, Checkout, Card, Cash, etc) knows nothing about each other and their internal states (the "child" part of the whole state you describe) get updated via messages, as per the Seneca pattern. 

So, I am suggesting there is no need for a "global" state or a "machine hierarchy" at all. Instead of passing state "down the chain" of hierarchy when something changes in one service, a message is sent by that service saying "This item was deleted" (with the relevant data eg SKU or whatever) and anyone who cares changes their state accordingly.

So, the whole checkout "state" is distributed amongst what are now "child" states but they don't know of each other's existence because each one is hidden inside the service. When they send / receive messages they are processing the message and have no idea who sent it or who they are sending to. 

That is the basis for the "microservice pattern". Mostly people talk about microservices server-side. I am interested in using them client-side as a way to architect the app in the browser (or other client-side app).

`sendParent` becomes irrelevant because there is no parent of any state. All states are self contained and have single "level" - they just send messages with data in order to communicate and update themselves and each other, but "anonymously".
 
Anyway, that is just how I think about it and when I build my first experiment it might well crash and burn. ;-)

**Comment by jonlaing** [comment]
Ah I see what you're saying. I'm not sure if that's the right pattern for what I'm doing. The hierarchy does offer some nice benefits with ensuring that certain machines are only listening for events at given times. As in, I don't want CreditCard to fire before Checkout has been invoked. I could add guards to prevent that, but the hierachical structure gives me that for free, so I think in this case the hierarchical pattern works well.

I think for now, I'll just force my custom store into my actions, plug it into react, and see how that works for me. If it works well, I'll come up with a more formal proposal for this project. I think I'll probably just do something like...

```typescript
// In some machine file
const actions = {
  updateShoppingCart: (ctx, evt) => Store.set({ shoppingCart: evt.data })
};

const services = {
  addToCart: (ctx) => addToCart({ shoppingCart: Store.get('shoppingCart'), item: ctx.selectedItem })
};

// in the UI
class App extends React.Component {
  state = {
    shoppingCart: null
  }

  componentDidMount() {
    Store.addListener(({ shoppingCart }) => this.setState({ shoppingCart }));
  }
  /* ... */
}
```

This proposal may end up being un-needed bloat for this project, but it would be nice to come up with a standard way of plugging in like that. Then again, I may have fundamentally misunderstood the intent behind this project (wouldn't be the first time), so I'll check back and see if there are more ideas floating around.

**Comment by flowt-au** [comment]
Cool. 
Just BTW: In the microservices pattern the messages can be "addressed" so the timing can be managed via a sequence of services, each receiving messages they are interested in then sending new messages when they have done their work so the chain of actions can stay in the correct sequence, as you say. :-) 

Anyway, good luck with your experiments!

**Comment by pyoner** [comment]
You can pass `store` from a parent machine context to the child machine using `data` property see https://xstate.js.org/docs/guides/communication.html#invoking-machines

**Comment by jonlaing** [comment]
@pyoner Yes, I suppose I could, and maybe that's the way I should go. I'd still have to make my custom store object, but I've already done that, so this could work too. And theoretically, it would be harder to screw that up than trying to sync disparate context values from different machines. Thanks for the suggestion.

**Comment by davidkpiano** [comment]
Just like @flowt-au is saying, imagine if each machine was a microservice that lived in a completely different machine. As with the Actor model, the goal is [eventual consistency](https://en.wikipedia.org/wiki/Eventual_consistency) -- not every "Actor" (machine instances in this case) will have the exact same "shared context" at the exact same time but that's okay, because they eventually will when they get "updates" via messages:

> State of actors is local and not shared, changes and data is propagated via messages, which maps to how modern memory hierarchy actually works. 
> 
> [From Akka docs](https://doc.akka.io/docs/akka/2.5/guide/actors-intro.html)

I think what you're trying to do, since you already know the pattern for doing this currently, is ~bypass~ simplify navigating the hierarchical levels for development convenience. This can _mostly_ be solved in userland; just think about the machine in isolation (pretend that other machines don't exist!) and come up with an event-based approach:

```js
// Contrived example
const childMachine = Machine({
  // ...
  on: {
    ...syncParent(['checkout', 'customer'])
  }
});
```

You can imagine `syncParent(...)` returning an object like:

```js
{
  "PARENT_CHANGE.checkout": {
    actions: assign({ checkout: (_, e) => e.value })
  },
  "PARENT_CHANGE.customer": {
    actions: assign({ customer: (_, e) => e.value })
  }
}
```

That's just one idea. I'll be exploring this more, as it seems to be a common pattern. We have to keep a few rules in mind though:

- There is no guarantee that each machine will have the same state at the same time (eventual consistency)
- Everything operates strictly as a state machine; "updates" are done via listening to events which carry updated data, because...
- Each machine's state is local and isolated
- Changes must propagate through the hierarchy (for determinism, concurrency, testability, etc.); i.e., we can't "skip" a parent.

**Comment by craigglennie** [comment]
> Just like @flowt-au is saying, imagine if each machine was a microservice that lived in a completely different machine. As with the Actor model, the goal is [eventual consistency](https://en.wikipedia.org/wiki/Eventual_consistency) -- not every "Actor" (machine instances in this case) will have the exact same "shared context" at the exact same time but that's okay, because they eventually will when they get "updates" via messages:

Probably I'm misunderstanding, but why is eventual consistency a goal? I've always thought of it as an unfortunate reality of certain types of systems (the kind where CAP theorem applies, for example) rather than something you actually want. I'd have thought it would be easier to reason about systems that aren't eventually consistent.

**Comment by davidkpiano** [comment]
> Probably I'm misunderstanding, but why is eventual consistency a goal?

Let's take the original example and apply it to a self-checkout machine:

![Self-checkout machine](https://upload.wikimedia.org/wikipedia/commons/thumb/5/51/IKEA_College_Park_self_checkout.jpg/220px-IKEA_College_Park_self_checkout.jpg)

The **Cart** (big computer) and the **Payment** (small computer to right) machines can be thought of as different services that live on different machines. These machines do _not_ share memory, and only share data by communicating with each other.

This is similar to, say, replacing the **Checkout** machine in the original example with e.g., a [Stripe checkout](https://stripe.com/billing), which obviously can't read your data, but needs to be _sent_ the order data and works asynchronously.

Eventual consistency is preferred when working with multiple async systems talking with each other, and makes each system highly testable and reliable in isolation. Strong consistency breaks down the more you introduce async data flow, regardless of if your services are on the same machine or not.

Just replace the "shared store" (in the original post) with a database or something async, and you can understand the problems that can arise with assuming that each services has the exact same data at the same time.

**Comment by jonlaing** [comment]
So, after doing a bit of looking around at how others use xstate, I think I'm overusing the `invoke` feature. I think my intent was to have one state machine to rule them all, but it seems like I'm better off having React invoke my various state machines, rather than having them invoke themselves. This offers a couple benefits in that my context doesn't grow crazily big, and I can actually inspect the state of a machine I call directly from React. I can use my "main" machine as a "store" of sorts to hang onto really important data that everything needs, like user info and whatever. Does this make sense to everyone?

Basically I'm thinking maybe the right way to do this is to decouple all of my machines from one another, and instead of invoking, call the specific machine I need for each screen as I need them from the UI itself.

**Comment by flowt-au** [comment]
Yep. That is the way I am approaching things. I will report back if I find an roadblocks when I starting building! Please let me know how it works out for you.

**Comment by davidkpiano** [comment]
> but it seems like I'm better off having React invoke my various state machines, rather than having them invoke themselves.

This is a good approach, and is essentially what is used in the TodoMVC example.

**Comment by jonlaing** [comment]
So, I'm thinkin that maybe this feature request isn't necessary if this is the proper approach. I've just about finished switching my app to an approach more similar to the TodoMVC example, and it's greatly simplified everything. Passing context back and forth isn't nearly as big of a problem for me now, and my machines are like half the code.

Also, I didn't realize that the `on` key can be used at the top level of a machine, which really changes the game for me.

So, yeah, I think the way things work now is probably fine for me, but maybe other people need this feature.

**Comment by davidkpiano** [comment]
> Also, I didn't realize that the `on` key can be used at the top level of a machine, which really changes the game for me.

I should definitely clarify this in the docs. Thanks for pointing that out 👏

**Comment by aislanmaia** [comment]
> I'm working on a pretty complex app with many invoked machines. After a while, it gets kind of hard to keep track of where you've properly passed down context from a parent machine into an invoked one. As the app has gotten more complex, tracking down these errors has become more of a slog.
> 
> As an example, let say I have an ecommerce app, and I need to keep my Order up to date. So I obviously have a browse machine, a cart machine, a checkout machine, a credit card checkout machine, a paypal checkout machine, etc etc. Not only do each one of those need to keep track of the Order, but each one of them can make changes to it.
> 
> This was a problem for React a while ago, where you'd have to pass down a single prop way down the structure, and it got unwieldy fast. Redux was one solution, but also there are React Contexts which are more lightweight.
> 
> I'm proposing something like this:
> 
> ```ts
> const BMachine = Machine({
>   /* ... */
>   states: {
>     c: {
>       onEntry: 'changeThing'
>     }
>   }
> }, {
>   actions: {
>     changeThing: assignStore({
>       thing: (ctx, evt, store) => store.thing + ' is the best'
>     })
>   }
> });
> 
> Machine({
>   /* ... */
>   states: {
>     a: {
>       invoke: {
>         src: 'b',
>         shareStore: true
>       }
>   }.
>   {
>     services: {
>         b: BMachine
>     }
>   }
> }).withStore({ thing: 'my value' });
> ```


So maybe with this store config we could integrate more tightly a Redux store out of the box ??

**Comment by jonlaing** [comment]
@aislanmaia I think there are already some efforts to do this on github, but none of them have seemed to gain much traction. That said, in another app I made, I just used the interpreter's `onDone()` method to fire off an action to Redux.

Example:
```javascript
interpreter.onDone(s => dispatch('SOME_ACTION', { data: s.currentState.context })
```

Or something like that... You can also do that with `onTransition()` and `onChange()` I believe. Each of them have different signatures, so you'll have to check because I don't remember them off the top of my head.

**Comment by aislanmaia** [comment]
@jonlaing so you've various machines but orchestrate them with a Main machine (kind a Parent of all), if I understood. My fear is that with this approach the state gets spread out between the machines, and having to synchronize them back to Main machine is a pain. That is the why my preference is for having a big one store centralized. If I change a part of state (data), this is automatically in sync with all the application, because all of them already are listening to the same local.

**Comment by davidkpiano** [comment]
> That is the why my preference is for having a big one store centralized.

Just a note, this can quickly become an antipattern. Performance can degrade quickly when the shared store has even moderate numbers of nested objects, and without proper change detection, machines listening to the store will have redundant transitions/updates.

There's also the issue of async persistence, which might not seem like an issue with a single client-side global store, but becomes one as soon as you need to persist any part of your store somewhere else.

Also, the idea of "update from anywhere" is dangerous, and makes code much less testable, because you can't predict where updates are coming from, or even when updates might happen. You can imagine a "global store" machine as a state machine with unguarded transitions for every piece of data it has. Makes it useless to analyze/visualize.

Furthermore, the intent behind events and clarity of logic are lost as well. Compare these two contrived examples:

**With global store**
```js
{
  on: {
    'STORE.CHANGE': [
      {
        actions: 'updateName',
        cond: (_, e) => {
          // Not even sure how you'd do change detection...
          return e.value.name !== e.previousValue.name
        }
      },
      {
        /* ... */
      }
    ]
  }
}
```

**Without global store**
```
{
  on: {
    'NAME.UPDATE': { actions: 'updateName' }
  }
}
```

Which one is more explicit? Which one is much easier to test? Which one is more clear, especially on a visualization?

**Comment by davidkpiano** [comment]
In short, if you want to use a global store, you already have one - it's the machine's `context` which can be accessed in any nested state within the machine.

If you separated your machine out into services communicating with each other, always make the assumption that those services can be async. That makes the notion of a global store less useful, since there needs to be a communication mechanism for sharing state.

**Comment by jonlaing** [comment]
@aislanmaia that was the way i was doing it before, but it quickly got unwieldy in exactly the way you said it might. So, I have the parent machine that has very basic storage for me, but most of the storage is in the machines themselves, and I pass down context via React, using `onDone` to tell the machine up the change to change if it needs to.

As an example:

```javascript
const Main = props => (
  <FSM machine={MainMachine}>
    {({state, context, send}) => (
      <Child
        someProp={context.someProp} 
        onDone={someProp => send({ type: 'updateSomeProp', someProp }) }
      />
    )}
  </FSM>
);

const Child = props => (
  <FSM machine={ChildMachine}>
    {({state, context, send}) => (
      <div>
        <button onClick={() => send('updateThing')}>Update</button>
        <button onClick={() => props.onDone(context.thing)}>Done</button>
      </div>
    )}
  </FSM>
);
```

`FSM` is just a custom component I made to handle spinning up and interpreter and updating its state based on the machine's state. The code is pretty simple so, I can make a gist if you want to see it, but that's how I'm going about things at the moment.

**Comment by aislanmaia** [comment]
> Just a note, this can quickly become an antipattern. Performance can degrade quickly when the shared store has even moderate numbers of nested objects, and without proper change detection, machines listening to the store will have redundant transitions/updates

Thats will not happen when we are connecting the components with Redux because we have fine grained control of what props the components needs to be renderized (in Elm world this is fine too). 

The problem you are stating here is more like when sharing all the state in the React Context, because it will force the nested objects (the components) to re-render unnecessarily when touching the shared state.

**Comment by aislanmaia** [comment]
> @aislanmaia that was the way i was doing it before, but it quickly got unwieldy in exactly the way you said it might. So, I have the parent machine that has very basic storage for me, but most of the storage is in the machines themselves, and I pass down context via React, using `onDone` to tell the machine up the change to change if it needs to.
> 
> As an example:
> 
> ```js
> const Main = props => (
>   <FSM machine={MainMachine}>
>     {({state, context, send}) => (
>       <Child
>         someProp={context.someProp} 
>         onDone={someProp => send({ type: 'updateSomeProp', someProp }) }
>       />
>     )}
>   </FSM>
> );
> 
> const Child = props => (
>   <FSM machine={ChildMachine}>
>     {({state, context, send}) => (
>       <div>
>         <button onClick={() => send('updateThing')}>Update</button>
>         <button onClick={() => props.onDone(context.thing)}>Done</button>
>       </div>
>     )}
>   </FSM>
> );
> ```
> 
> `FSM` is just a custom component I made to handle spinning up and interpreter and updating its state based on the machine's state. The code is pretty simple so, I can make a gist if you want to see it, but that's how I'm going about things at the moment.

I'm really interested to see this thing. One question thats up right the moment I see you using the React Context is exactly about the unnecessarily rerenders. I've just make some attempts using React Context and passing the machine, but I noticed that the parent and child components were rerendering more than they should.

**Comment by davidkpiano** [comment]
Related: #424

**Comment by davidkpiano** [comment]
The latest [TodoMVC example](https://xstate.js.org/docs/examples/todomvc.html) now reflects how this can be accomplished with Actors as of version 4.6.

I'll make more detailed examples 🔜.

**Comment by gilesbutler** [comment]
> > but it seems like I'm better off having React invoke my various state machines, rather than having them invoke themselves.
> 
> This is a good approach, and is essentially what is used in the TodoMVC example.

Is this still the case @davidkpiano? I'm a bit confused from this thread. I've created a Vue project that has an app machine and App component. The app machine invokes child machines as the user navigates to different routes throughout the app. I then pass those invoked services as props to the child vue components from the parent App component. The child components then call `useService` and pass in the already running service that has been passed down to them. The running service may invoke other machines and pass those services down to other child components. From what I'm reading in this thread though that isn't the way to go? Rather than passing the invoked services down from parent to child, should the child components be calling `useMachine` instead and invoke their own services?

**Comment by davidkpiano** [comment]
> Rather than passing the invoked services down from parent to child, should the child components be calling `useMachine` instead and invoke their own services?

No, your initial approach is correct. You can/should pass down those services as props, and interact with them using `useService` (and soon, `useActor`, which isn't in `@xstate/vue` yet).

**Comment by gilesbutler** [comment]
Thanks for clearing that up @davidkpiano.

How will `useActor` differ from `useService`?

**Comment by davidkpiano** [comment]
> How will `useActor` differ from `useService`?

There is little difference. In fact, the `useService()` hook creates an actor from the `service` and then uses `useActor()`:

```ts
// ...

export function useService<
  TContext,
  TEvent extends EventObject,
  TTypestate extends Typestate<TContext> = { value: any; context: TContext }
>(
  service: Interpreter<TContext, any, TEvent, TTypestate>
): [State<TContext, TEvent, any, TTypestate>, Sender<TEvent>] {
  const serviceActor = useMemo(() => fromService(service), [service]);

  return useActor<TEvent, State<TContext, TEvent, any, TTypestate>>(
    serviceActor,
    (actor) => (actor as typeof serviceActor).current
  );
}
```

**Comment by davidkpiano** [comment]
> How will `useActor` differ from `useService`?

There is little difference; `useService` converts the service to an actor, then passes that to `useActor`.

**Comment by justSteve** [comment]
> Yep. That is the way I am approaching things. I will report back if I find an roadblocks when I starting building! Please let me know how it works out for you.

I would be intensely interested in a '_2-years-hence_' summary on lessons learned for the above. Preceedingly you've referenced not using xstate. Have you started with it in the time since?

**Comment by zacharyweidenbach** [comment]
This was an interesting thread and touches on some of the main pain points I'm unsure how to resolve. In my particular case, I am using react-native + react-navigation. The main issue is that react-navigation keeps screens mounted after navigating forward in the same navigation stack. If machines are organized at the screen level, and some mutation occurs in a later screen that needs to cascade to the previous screen (which is still mounted) to reflect the change, it's unclear the best way to do that.

In this contrived example, let's say there are 2 screens:
1. ItemListScreen
2. ItemEditScreen
Let's also say you can only get to ItemEditScreen through ItemListScreen

The ItemListScreen fetches a list of items. The ItemEditScreen will edit a single item via PATCH request to /items/:id
Maybe you can see the problem already. If the ItemEditScreen mutates the data and we want that reflected on the list on the previous screen (which is still mounted btw), we need to manage that somehow.

One approach is to update the context of the ItemListScreen machine as a side effect of the EDIT event on the ItemEditScreen. But this doesn't feel very elegant or intuitive to follow. And not very scalable if the side effect should cascade over many previous screens (of which it could be hard/impossible to know which may still be mounted).

Alternatively, the fetching and caching of the api data could be its own machine living in react context that both screens share, and the screens can trigger the fetch/cache machine from within their own respective screens, and similarly the EditItemScreen machine can trigger a cache invalidation that will be detected by the ItemListScreen and it will respond by triggering a refetch from the fetch/cache machine. But this is beginning to feel very Rube Goldberg and hard to follow for what is simple api cache invalidation.

Perhaps this is more a problem with react-navigation than anything else. I'm pretty sure in the world of react web apps using react-router, components just unmount when they are navigated away from, so all of this becomes moot because the machines on each screen will just run through their bootstrap step and get the most up-to-date version of the data. It might be that this unfortunate design detail of react-navigation means that a machine (propagated via react context wrapping all screens in question) must take ownership of all screens that need to share state, but that's a big ol' bummer of a pattern.

The other (probably goofball) solution I could think of is to have the api cache data live as singleton functions that expose out Observables and a handful of methods to change the data in the cache (and trigger observable events when called). Then the screen machines that rely on the cache data will have parallel states, where in one sibling state, it invokes the observable store to begin listening for events (like cache invalidation), and in the other sibling state, it does all the other things it needs to do to manage the state of the screen. Then, if another screen machine mounts and needs to invalidate the cache that another machine might be reliant on, the observable will propagate that change. This also feels pretty hard to follow, and the mechanics of subscribing to the store by invoking an observable seems clunky.

Maybe the simplest answer is to try and listen for when the app navigates "back" to a screen that's stayed mounted (and is now in 'focus' again) in react, and send a REFRESH event that triggers some workflow that discovers the cache is now empty and it needs to fetch the list. This would keep the machines isolated from each other's side effects, and basically eliminates the need for a reactive global store altogether. But there might be a brief moment when the machine is trying to detect if it should refetch and the user might catch a glimpse of the old data they thought they just edited.

I'll start with that last one, but I worry it doesn't address scenarios where a truly global reactive storage is necessary.

**Comment by Andarist** [comment]
I think that, in this case, some kind of a caching layer is desirable for the best UX anyway. It's not really your screen that manages data - it just needs to display and edit it (and perhaps kick off the initial network request but that, ideally, should happen as soon as you know that you will need a particular data, so not really during rendering). 

The situation described here isn't that different from a listing page with the edit modal - although I guess in such a situation usually you have a parent-child relationship so it's easier to share data and put the whole data ownership in the listing component itself. Whereas, if I understand correctly, you don't have that luxury with react-navigation as those are completely different components, it just happens that old screens are not always fully unmounted.

Note that you are also not limited to only using `useMachine` (that ties the machine to the lifecycle of the component) - but you can also just create a "singleton" service using "raw" `interpret` API and consume it in your components using `useActor`/`useSelector`. This might make it easier to keep this data management in a single place and decouple it from components
